amy/consents/admin.py

```diff
@@ -1,15 +1,71 @@
+from django.contrib import messages
+from django.contrib.admin.options import csrf_protect_m
+from django.http.response import HttpResponseRedirect
+from urllib.parse import unquote
+
 from consents.models import Consent
 from consents.models import Term, TermOption
 from django.contrib import admin
 
 
+class ArchiveActionMixin:
+    """
+    Used for AdminModels that have the CreatedUpdatedArchivedMixin
+    and need to archive rather than delete the model.
+    """
+
+    change_form_template = "consents/admin_change_form_term.html"
+    actions = ["archive"]
+    readonly_fields = ("archived_at",)
+
+    @csrf_protect_m
+    def changeform_view(self, request, object_id=None, form_url="", extra_context=None):
+        if request.method == "POST" and "_archive" in request.POST:
+            obj = self.get_object(request, unquote(object_id))
+            self.archive_single(request, obj)
+            return HttpResponseRedirect(request.get_full_path())
+
+        return admin.ModelAdmin.changeform_view(
+            self,
+            request,
+            object_id=object_id,
+            form_url=form_url,
+            extra_context=extra_context,
+        )
+
+    def archive_single(self, request, obj):
+        if obj.archived_at is not None:
+            self.message_user(
+                request,
+                f"Error: Cannot archive. {obj} is already archived.",
+                level=messages.ERROR,
+            )
+        else:
+            obj.archive()
+
+    def has_delete_permission(self, *args, **kwargs):
+        return False
```

Comment:
What's the purpose of this method? Can you document in a docstring? Edit: or comment

Reply:
I'll add a comment with a link to the documentation. This is a feature of django admin, that lets django know that this admin view cannot delete models. Admin views can typically delete models by default. I instead want them archived from the admin view and related models archived.